messagebus/src/vespa/messagebus/network/rpcsendv2.cpp

```diff
@@ -0,0 +1,247 @@
+            (respons == METHOD_RETURN);
+}
+
+void
+RPCSendV2::build(FRT_ReflectionBuilder & builder)
+{
+    builder.DefineMethod(METHOD_NAME, METHOD_PARAMS, METHOD_RETURN, true, FRT_METHOD(RPCSendV2::invoke), this);
+    builder.MethodDesc("Send a message bus slime request and get a reply back.");
+    builder.ParamDesc("encoding", "0=raw, 6=lz4");
+    builder.ParamDesc("uncompressedBlobSize", "Uncompressed blob size");
+    builder.ParamDesc("message", "The message blob in slime");
+    builder.ReturnDesc("encoding",  "0=raw, 6=lz4");
+    builder.ReturnDesc("uncompressedBlobSize", "Uncompressed blob size");
+    builder.ReturnDesc("reply", "The reply blob in slime.");
+}
+
+const char *
+RPCSendV2::getReturnSpec() const {
+    return METHOD_RETURN;
+}
+
+namespace {
+class OutputBuf : public vespalib::Output {
+public:
+    OutputBuf(size_t estimatedSize) : _buf(estimatedSize) { }
+    DataBuffer & getBuf() { return _buf; }
+private:
+    vespalib::WritableMemory reserve(size_t bytes) override {
+        _buf.ensureFree(bytes);
+        return vespalib::WritableMemory(_buf.getFree(), _buf.getFreeLen());
+    }
+    Output &commit(size_t bytes) override {
+        _buf.moveFreeToData(bytes);
+        return *this;
+    }
+    DataBuffer _buf;
+};
+}
+
+void
+RPCSendV2::encodeRequest(FRT_RPCRequest &req, const Version &version, const Route & route,
+                         const RPCServiceAddress & address, const Message & msg, uint32_t traceLevel,
+                         const PayLoadFiller &filler, uint64_t timeRemaining) const
+{
+    Slime slime;
+    Cursor & root = slime.setObject();
+
+    root.setString(VERSION_F, version.toString());
+    root.setString(ROUTE_F, route.toString());
+    root.setString(SESSION_F, address.getSessionName());
+    root.setBool(USERETRY_F, msg.getRetryEnabled());
+    root.setLong(RETRY_F, msg.getRetry());
+    root.setLong(TIMELEFT_F, timeRemaining);
+    root.setString(PROTOCOL_F, msg.getProtocol());
+    root.setLong(TRACELEVEL_F, traceLevel);
+    filler.fill(BLOB_F, root);
+
+    OutputBuf rBuf(8192);
+    BinaryFormat::encode(slime, rBuf);
+    ConstBufferRef toCompress(rBuf.getBuf().getData(), rBuf.getBuf().getDataLen());
+    DataBuffer buf(vespalib::roundUp2inN(rBuf.getBuf().getDataLen()));
+    CompressionConfig::Type type = compress(_net->getCompressionConfig(), toCompress, buf, false);
+
+    FRT_Values &args = *req.GetParams();
+    req.SetMethodName(METHOD_NAME);
+    args.AddInt8(type);
+    args.AddInt32(toCompress.size());
+    args.AddData(buf.stealBuffer(), buf.getDataLen());
+}
+
+namespace {
+
+class ParamsV2 : public RPCSend::Params
+{
+public:
+    ParamsV2(const FRT_Values &arg)
+        : _slime()
+    {
+        uint8_t encoding = arg[0]._intval8;
+        uint32_t uncompressedSize = arg[1]._intval32;
+        DataBuffer uncompressed(arg[2]._data._buf, arg[2]._data._len);
+        ConstBufferRef blob(arg[2]._data._buf, arg[2]._data._len);
```

Comment:
Dangerous to trust buffer lengths received from the network. Should check lengths against received blob size.

Reply:
Ignore comment, did not originally see that length is from RPC layer itself 🦄 👌